src/mlpack/core/tree/spill_tree/spill_single_tree_traverser_impl.hpp

```diff
@@ -52,7 +52,14 @@ SpillSingleTreeTraverser<RuleType, Defeatist>::Traverse(
   if (referenceNode.IsLeaf())
   {
     for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
-      rule.BaseCase(queryIndex, referenceNode.Point(i));
+    {
+      size_t referenceIndex = referenceNode.Point(i);
+
+      // Don't run the basecase for the reference point if it is already present in
+      // the neighbor list of the querypoint.
+      if(!rule.InNeighborList(queryIndex, referenceIndex))
+        rule.BaseCase(queryIndex, referenceIndex);
```

Comment:
> Actually it seems like this modification could cause the `BaseCase()` to run multiple times for a given `queryIndex` and `referenceIndex`. This could happen for kNN if the `referenceIndex` isn't in the top k neighbors for a given `queryIndex`.

I am not able to understand how that would happen. Can you provide an example.

Reply:
A spill tree may have overlapping nodes; so, basically, a left child and a right child of some node may "share" some points.  That is, there could exist `i` and `j` such that `leftChild.Point(i) == rightChild.Point(j)`.  In this situation, we could call `rule.BaseCase()` a second time for a given `queryIndex`/`referenceIndex` pair.

However, I'm realizing now that actually this situation was also possible *before* your change---which I suppose is a bug in the spill tree implementation!  Do you want to handle that as part of this PR, or should I open another issue?  I could probably produce a failing test case for it. :)

Reply:
Yeah, I can work on it too.

Reply:
I'll see if I can write up a test case today or tomorrow. :+1:

Reply:
I dug into this today and it reminded me of some discussions from years ago.  First of all, it is true that the spill tree's single tree traverser *does* have the problem that when not in `Defeatist` mode, it may call the base case multiple times for certain points.  However, using the spill tree in non-defeatist mode is actually just a worse version of using a kd-tree (in most situations).

So, this non-defeatist mode is never actually used, and this is intentional.  If you take a look at `src/mlpack/methods/ns_model.hpp`, the type used is `SpillKNN`, which uses defeatist search.  That is, when I run the command-line program `mlpack_knn -t spill -a single_tree ...`, this actually uses defeatist search and is identical to `-a greedy`, actually.

Overall I think it is fine to leave this problem in, but, it might be helpful to add a warning in the constructor like

```
if (!Defeatist)
  Log::Warn << "Non-defeatist search with spill trees may cause duplicate results!" << std::endl;
```

or something.

Anyway, I realize I've written a lot here and what I was hoping to do was add some context to the discussion.  If it doesn't make sense though, or I didn't do a good job explaining it, the summary is that I don't think I should have left the original comment for this discussion to begin with; I think it's out of scope for this issue.  If you like, we can add that warning I suggested or that could be handled some other time. :+1: